translations_parser/parser.py

```
@@ -0,0 +1,228 @@
+    r"Time[ :]+(?P<time>[\d\.]+s)[ :]+"
+    r"(?P<rate>[\d\.]+ words\/s)[ :]+"
+    r"gNorm[ :]+(?P<gnorm>[\d\.]+)"
+)
+
+
+@dataclass
+class Training:
+    epoch: str
+    up: str
+    sen: str
+    cost: str
+    time: int
+    rate: str
+    gnorm: str
+
+
+@dataclass
+class Validation:
+    epoch: str
+    up: str
+    chrf: str
+    ce_mean_words: str
+    bleu_detok: str
+
+
+@dataclass
+class TrainingLog:
+    """Results from the parsing of a training log file"""
+
+    # Marian information
+    info: dict
+    # Runtime configuration
+    configuration: dict
+    training: List[Training]
+    validation: List[Validation]
+    # Dict of log lines indexed by their header (e.g. marian, data, memory)
+    logs: dict
+
+
+class TrainingParser:
+    def __init__(self, logs_iter):
+        # Iterable reading logs lines
+        self.logs_iter = logs_iter
+        self.parsed = False
+        self.info = {}
+        self.config = {}
+        self.training = []
+        # Dict mapping (epoch, up) to values parsed on multiple lines
+        self.validation_entries = defaultdict(dict)
+        self.indexed_logs = defaultdict(list)
+
+    def get_headers(self, line):
+        """
+        Returns a list of tuples representing the headers of a log line
+        and the position of the last index
+        """
+        matches = list(HEADER_RE.finditer(line))
+        if not matches:
+            return ((), None)
+        return ([tuple(m.group("value").split()) for m in matches], matches[-1].span()[-1])
+
+    def check_task_timestamp_header(self, values):
+        """
+        Check a header value matching ('task', <timestamp>)
+        and return the deduced timestamp
+        """
+        if not values or len(values) != 2:
+            return
+        base, timestamp = values
+        if base != "task":
+            return
+        try:
+            return datetime.fromisoformat(timestamp.rstrip("Z"))
+        except ValueError:
+            return
```

This should raise, an invalid date from TC is a bad thing